### PR TITLE
Fix #1713 RelevanceConfig\Structure\Data parent class

### LINKED ISSUE #1713: Wrong virtual type declaration
Following virtual type shouldn't extend Magento\Config\Model\Config\Structure\Data but Magento\Framework\Config\Data\Scoped

```
    <virtualType name="Smile\ElasticsuiteCore\Model\Search\Request\RelevanceConfig\Structure\Data" type="Magento\Config\Model\Config\Structure\Data">
        <arguments>
            <argument name="reader" xsi:type="object">Smile\ElasticsuiteCore\Search\Request\RelevanceConfig\Structure\Reader</argument>
            <argument name="cacheId" xsi:type="string">smile_elasticsuite_relevance_configuration_structure</argument>
        </arguments>
    </virtualType>
```

Indeed Magento\Config\Model\Config\Structure\Data is concrete implementation for magento system config.
Please use the generic one: Magento\Framework\Config\Data\Scoped

### PR DESCRIPTION
**Pull request by thomas-kl1**
See: https://github.com/Smile-SA/elasticsuite/issues/1713

### CODE REVIEW

File: src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+declare(strict_types=1);
+
+namespace Smile\ElasticsuiteCore\Model\Search\Request\RelevanceConfig\Structure;
+
+use Magento\Framework\Config\Data\Scoped;
+
+class Data extends Scoped
+{
+    public function merge(array $config): void
```
Comment (rbayet) [inline src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: Please add PHPdoc for the function.
You can use 
```
    /**
     * {@inheritDoc}
     */
```

File: src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+declare(strict_types=1);
+
+namespace Smile\ElasticsuiteCore\Model\Search\Request\RelevanceConfig\Structure;
+
+use Magento\Framework\Config\Data\Scoped;
+
+class Data extends Scoped
```
Comment (rbayet) [inline src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: Please add PHPdoc for the class with the following template 
```
/**
 * Description goes here.
 *
 * @category Smile
 * @package  Smile\ElasticsuiteCore
 * @author   Thomas Klein <thomasklein876@gmail.com>
 */
```

File: src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * Do not edit or add to this file if you wish to upgrade Smile ElasticSuite to newer
+ * versions in the future.
+ *
+ * @category  Smile
+ * @package   Smile\ElasticsuiteCore
+ * @author    Thomas Klein <thomasklein876@gmail.com>
+ * @copyright 2020 Smile
+ * @license   Open Software License ("OSL") v. 3.0
+ */
+declare(strict_types=1);
```
Comment (rbayet) [inline src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: Please put spaces around the assign operator ->
`declare(strict_types = 1);`

File: src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php (hunk 1 above)
Comment (thomas-kl1) [reply src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: You should update your phpcs rules, brackets over @inheritDoc shouldn't be used and this tag has only interest if the description or arguments is updated in the overridden method.
The method is self descriptive and already signed.

File: src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php (hunk 2 above)
Comment (thomas-kl1) [reply src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: Will do

File: src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php (hunk 3 above)
Comment (thomas-kl1) [reply src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: Will do

File: src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php (hunk 1 above)
Comment (rbayet) [reply src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: Maybe. But we use extensively {@inheritDoc} as it is and the rule about every function being documented is not going away anytime soon.

File: src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * @author   Thomas Klein <thomasklein876@gmail.com>
+ */
+class Data extends Scoped
+{
+    public function merge(array $config): void
+    {
+        parent::merge($config['config']['system'] ?? $config);
+    }
+}
+
```
Comment (rbayet) [inline src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: Can you also check the end of your file ?
Travis is rambling about "Expected 1 blank line at end of file; 2 found".

Regards,

Comment (thomas-kl1) [reply src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: whups, edited with gedit, my bad

File: src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * @category Smile
+ * @package  Smile\ElasticsuiteCore
+ * @author   Thomas Klein <thomasklein876@gmail.com>
+ */
+class Data extends Scoped
+{
+    /**
+     * Merge default config data with extended settings
+     *
+     * @param array $config
```
Comment (rbayet) [inline src/module-elasticsuite-core/Model/Search/Request/RelevanceConfig/Structure/Data.php]: Almost there (sorry :|), could you add a comment after "$config" and have an empty comment line between [@param ](url)and `@return` ?
see for instance https://github.com/Smile-SA/elasticsuite/blob/67fbc17c2bfbf9383e654b454e50e8cbedf4643a/src/module-elasticsuite-core/Index/Mapping/Field.php#L278

Regards,

**Comment by thomas-kl1** [comment]
Hi, any ETA?

**Comment by rbayet** [comment]
> Hi, any ETA?

Hello @thomas-kl1,

I'm going to check that they are no functional impact. To be honest, both @romainruaud and I are a bit dumbfunded by the `merge` function.
I also want to check when Magento\Framework\Config\Data\Scoped was introduced.

Regards,

**Comment by rbayet** [comment]
> To be honest, both @romainruaud and I are a bit dumbfunded by the `merge` function.

OK, it's simply what `\Magento\Config\Model\Config\Structure\Data` does.

> I also want to check when Magento\Framework\Config\Data\Scoped was introduced.

It was there from the beginning, but like `\Magento\Config\Model\Config\Structure\Data`, it was not tagged as @api until 05/2017.

**Comment by thomas-kl1** [comment]
"merge" is for ignoring the two nodes config/system in the elasticsuite_relevance.xml file. It's the same because you copied/pasted how native config works in Magento.

**Comment by rbayet** [comment]
The PR fails both in developer and production modes on M2 2.3.4 (OpenSource and Commerce versions).

Here is the exception trace in developer mode : 
```
1 exception(s):
Exception #0 (Magento\Framework\Exception\RuntimeException): Type Error occurred when creating object: Magento\Config\Model\Config\Structure\Interceptor

Exception #0 (Magento\Framework\Exception\RuntimeException): Type Error occurred when creating object: Magento\Config\Model\Config\Structure\Interceptor
<pre>#1 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create() called at [vendor/magento/framework/ObjectManager/ObjectManager.php:56]
#2 Magento\Framework\ObjectManager\ObjectManager->create() called at [vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:169]
#3 Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument() called at [vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:273]
#4 Magento\Framework\ObjectManager\Factory\AbstractFactory->getResolvedArgument() called at [vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:236]
#5 Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime() called at [vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:34]
#6 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments() called at [vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php:59]
#7 Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create() called at [vendor/magento/framework/ObjectManager/ObjectManager.php:56]
#8 Magento\Framework\ObjectManager\ObjectManager->create() called at [vendor/magento/framework/App/ActionFactory.php:44]
#9 Magento\Framework\App\ActionFactory->create() called at [vendor/magento/framework/App/Router/Base.php:306]
#10 Magento\Framework\App\Router\Base->matchAction() called at [vendor/magento/framework/App/Router/Base.php:167]
#11 Magento\Framework\App\Router\Base->match() called at [vendor/magento/framework/App/FrontController.php:95]
#12 Magento\Framework\App\FrontController->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#13 Magento\Framework\App\FrontController\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#14 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#15 Magento\Framework\App\FrontController\Interceptor->___callPlugins() called at [generated/code/Magento/Framework/App/FrontController/Interceptor.php:26]
#16 Magento\Framework\App\FrontController\Interceptor->dispatch() called at [vendor/magento/framework/App/Http.php:116]
#17 Magento\Framework\App\Http->launch() called at [generated/code/Magento/Framework/App/Http/Interceptor.php:24]
#18 Magento\Framework\App\Http\Interceptor->launch() called at [vendor/magento/framework/App/Bootstrap.php:261]
#19 Magento\Framework\App\Bootstrap->run() called at [index.php:39]
</pre>
```

**Pull request closed without merging**